parlai/core/worlds.py

```diff
@@ -1499,6 +1531,629 @@ def shutdown(self):
         self.inner_world.shutdown()
 
 
+class QueueSignal(Enum):
+    """
+    Command signal used in the Queues for QueueWorld.
+
+    BATCH:                normal result; a batch is in the Queue
+    WORKER_FINISHED:      a worker is finished sending batch
+    RESET:                tell a worker to reset
+    RESET_METRICS:        tell a worker to reset_metrics
+    REPORT:               tell a worker to report
+    BEGIN:                tell a worker to begin again (valid worlds)
+    TERMINATE:            tell a worker to terminate
+    """
+
+    BATCH = auto()
+    WORKER_FINISHED = auto()
+    RESET = auto()
+    RESET_METRICS = auto()
+    REPORT = auto()
+    BEGIN = auto()
+    TERMINATE = auto()
+
+
+class PWorldProcess(Process):
+    """
+    Process that runs a PWorld.
+    """
+
+    def __init__(
+        self,
+        opt: Opt = None,
+        world: Union[DialogPartnerWorld, MultiWorld] = None,
+        pworld_class: type = None,
+        produce_queue: Queue = None,
+        report_queue: Queue = None,
+        consume_queue: Queue = None,
+        worker_idx: int = None,
+    ):
+        super().__init__(daemon=True)
+        self.p_world: PWorld = pworld_class(
+            opt, world, produce_queue, report_queue, consume_queue, worker_idx
+        )
+        self.produce_queue = produce_queue
+        self.consume_queue = consume_queue
+        self.worker_idx = worker_idx
+
+    def run(self):
+        """
+        Run a PWorld to completion.
+
+        The function first runs a world until `epoch_done()` is True, and then awaits
+        further instruction
+        """
+        while True:
+            while not self.p_world.epoch_done():
+                self.p_world.parley()
+            # one more time to observe
+            self.produce_queue.put((QueueSignal.WORKER_FINISHED, None, self.worker_idx))
+            while True:
+                signal, batch = self.consume_queue.get()
+                if signal == QueueSignal.BATCH:
+                    # TODO: determine why this happens
```

Comment:
what is this TODO?

Reply:
still was not sure how I could reach this, will continue to test a little bit